stable_baselines3/common/recurrent/torch_layers.py

```diff
@@ -0,0 +1,205 @@
+import abc
+from typing import Any, Dict, Generic, Optional, Tuple, TypeVar
+
+import gymnasium as gym
+import torch as th
+
+from stable_baselines3.common.preprocessing import get_flattened_obs_dim
+from stable_baselines3.common.pytree_dataclass import TensorTree, tree_flatten, tree_map
+from stable_baselines3.common.recurrent.type_aliases import (
+    GRURecurrentState,
+    LSTMRecurrentState,
+)
+from stable_baselines3.common.torch_layers import (
+    BaseFeaturesExtractor,
+    CombinedExtractor,
+    FlattenExtractor,
+    NatureCNN,
+)
+from stable_baselines3.common.type_aliases import TorchGymObs
+
+RecurrentState = TypeVar("RecurrentState", bound=TensorTree)
+
+RecurrentSubState = TypeVar("RecurrentSubState", bound=TensorTree)
+
+ExtractorInput = TypeVar("ExtractorInput", bound=TorchGymObs)
+
+
+class RecurrentFeaturesExtractor(BaseFeaturesExtractor, abc.ABC, Generic[ExtractorInput, RecurrentState]):
+    @abc.abstractmethod
+    def recurrent_initial_state(
+        self, n_envs: Optional[int] = None, *, device: Optional[th.device | str] = None
+    ) -> RecurrentState:
+        ...
+
+    @abc.abstractmethod
+    def forward(
+        self, observations: ExtractorInput, state: RecurrentState, episode_starts: th.Tensor
+    ) -> Tuple[th.Tensor, RecurrentState]:
+        ...
+
+    @staticmethod
+    def _process_sequence(
+        rnn: th.nn.RNNBase, inputs: th.Tensor, init_state: RecurrentSubState, episode_starts: th.Tensor
+    ) -> Tuple[th.Tensor, RecurrentSubState]:
+        (state_example, *_), _ = tree_flatten(init_state, is_leaf=None)
+        n_layers, batch_sz, *_ = state_example.shape
+
+        # Batch to sequence
+        # (padded batch size, features_dim) -> (n_seq, max length, features_dim) -> (max length, n_seq, features_dim)
+        seq_len = inputs.shape[0] // batch_sz
+        seq_inputs = inputs.view((batch_sz, seq_len, *inputs.shape[1:])).swapaxes(0, 1)
+        episode_starts = episode_starts.view((batch_sz, seq_len)).swapaxes(0, 1)
+
+        if th.any(episode_starts[1:]):
+            raise NotImplementedError("Resetting state in the middle of a sequence is not supported")
+
+        first_state_is_not_reset = (~episode_starts[0]).contiguous()
+        # Shape here is (n_layers, batch_sz)
```

Comment:
This comment is helpful, but it could also be an assert (and I trust asserts more).

Reply:
Thanks, done! 53d2f2e1 

I think the original comment was kind of confusing and didn't give a full view.